File: lib/python/opentoken/src/test/opentoken/attributes/person/postal_code_attribute_test.py
# src/test/opentoken/attributes/person/test_postal_code_attribute.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import pickle
import pytest
import threading
from concurrent.futures import ThreadPoolExecutor, as_completed

from opentoken.attributes.person.postal_code_attribute import PostalCodeAttribute


class TestPostalCodeAttribute:
    """Test cases for PostalCodeAttribute class."""
    
    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.postal_code_attribute = PostalCodeAttribute()
    
    def test_get_name_should_return_postal_code(self):
        """Test that get_name returns 'PostalCode'."""
        assert self.postal_code_attribute.get_name() == "PostalCode"
    
    def test_get_aliases_should_return_postal_code_and_zip_code(self):
        """Test that get_aliases returns PostalCode and ZipCode."""
        expected_aliases = ["PostalCode", "ZipCode"]
        actual_aliases = self.postal_code_attribute.get_aliases()
        assert actual_aliases == expected_aliases
    
    def test_normalize_should_return_first_5_digits(self):
        """Test normalization returns first 5 digits for US ZIP codes."""
        assert self.postal_code_attribute.normalize("10001-6789") == "10001"
        assert self.postal_code_attribute.normalize("10001") == "10001"
    
    def test_normalize_should_handle_canadian_postal_codes(self):
        """Test normalization of Canadian postal codes."""
        assert self.postal_code_attribute.normalize("K1A0A6") == "K1A 0A6"
        assert self.postal_code_attribute.normalize("k1a0a6") == "K1A 0A6"
        assert self.postal_code_attribute.normalize("K1A 0A6") == "K1A 0A6"
        assert self.postal_code_attribute.normalize("m5v3l9") == "M5V 3L9"
        assert self.postal_code_attribute.normalize("H3Z2Y7") == "H3Z 2Y7"
        assert self.postal_code_attribute.normalize("t2x1v4") == "T2X 1V4"
    
    def test_validate_should_return_true_for_valid_postal_codes(self):
        """Test validation for valid US postal codes."""
        assert self.postal_code_attribute.validate("95123 ") is True
        assert self.postal_code_attribute.validate(" 95123") is True
        assert self.postal_code_attribute.validate("95123") is True
        assert self.postal_code_attribute.validate("95123-6789") is True
        assert self.postal_code_attribute.validate("65201-6789") is True
    
    def test_validate_should_return_true_for_valid_canadian_postal_codes(self):
        """Test validation for valid Canadian postal codes."""
        assert self.postal_code_attribute.validate("K1A 0A7") is True
        assert self.postal_code_attribute.validate("K1A0A7") is True
        assert self.postal_code_attribute.validate("k1a 0a7") is True
        assert self.postal_code_attribute.validate("k1a0a7") is True
        assert self.postal_code_attribute.validate("M5V 3L9") is True
        assert self.postal_code_attribute.validate("H3Z 2Y7") is True
        assert self.postal_code_attribute.validate("T2X 1V4") is True
        assert self.postal_code_attribute.validate(" K1A 0A7 ") is True
        assert self.postal_code_attribute.validate("  K1A0A7  ") is True
    
    def test_normalize_should_handle_whitespace(self):
        """Test different types of whitespace handling."""
        attribute = PostalCodeAttribute()
        
        # Test different types of whitespace
        assert attribute.normalize("10001") == "10001", "No whitespace"
        assert attribute.normalize(" 10001") == "10001", "Leading space"
        assert attribute.normalize("10001 ") == "10001", "Trailing space"
        assert attribute.normalize(" 10001 ") == "10001", "Leading and trailing spaces"
        assert attribute.normalize("  10   001  ") == "10   001", "Multiple spaces"
    
    def test_normalize_should_not_handle_inner_whitespace(self):
        """Test that inner whitespace is preserved."""
        attribute = PostalCodeAttribute()
        
        # Test inner whitespace
        assert attribute.normalize("1 0 0 0 1") == "1 0 0 0 1", "Spaces between digits"
        assert attribute.normalize("10\t001") == "10\t001", "Tab character"
        assert attribute.normalize("10\n001") == "10\n001", "Newline character"
        assert attribute.normalize("10\r\n001") == "10\r\n001", "Carriage return and newline"
        assert attribute.normalize("K1A  0A7") == "K1A  0A7", \
               "Inner space should not be normalized for Canadian postal code"
    
    def test_validate_should_return_false_for_invalid_postal_codes(self):
        """Test validation for invalid postal codes."""
        assert self.postal_code_attribute.validate(None) is False, \
               "Null value should not be allowed"
        assert self.postal_code_attribute.validate("") is False, \
               "Empty value should not be allowed"
        assert self.postal_code_attribute.validate("1234") is False, \
               "Short postal code should not be allowed"
        assert self.postal_code_attribute.validate("12345") is False, \
               "Invalid postal code should not be allowed"
        assert self.postal_code_attribute.validate("54321") is False, \
               "Invalid postal code should not be allowed"
        assert self.postal_code_attribute.validate("123456") is False, \
               "Long postal code should not be allowed"
        assert self.postal_code_attribute.validate("1234-5678") is False, \
               "Invalid format should not be allowed"
        assert self.postal_code_attribute.validate("abcde") is False, \
               "Non-numeric should not be allowed"
        
        # Invalid Canadian postal code formats
        assert self.postal_code_attribute.validate("K1A") is False, \
               "Incomplete Canadian postal code should not be allowed"
        assert self.postal_code_attribute.validate("K1A 0A") is False, \
               "Incomplete Canadian postal code should not be allowed"
        assert self.postal_code_attribute.validate("K1A 0A67") is False, \
               "Too long Canadian postal code should not be allowed"
        assert self.postal_code_attribute.validate("K11 0A6") is False, \
               "Invalid Canadian postal code format should not be allowed"
        assert self.postal_code_attribute.validate("KAA 0A6") is False, \
               "Invalid Canadian postal code format should not be allowed"
        assert self.postal_code_attribute.validate("K1A 0AA") is False, \
               "Invalid Canadian postal code format should not be allowed"
    
    def test_normalize_thread_safety(self):
        """Test thread safety of normalize method."""
        thread_count = 100
        test_postal_code = "10001-6789"
        expected_result = "10001"
        results = []
        
        def normalize_postal_code():
            """Function to be executed by each thread."""
            try:
                result = self.postal_code_attribute.normalize(test_postal_code)
                return result
            except Exception as e:
                pytest.fail(f"Thread failed with exception: {e}")
        
        # Use ThreadPoolExecutor for better thread management
        with ThreadPoolExecutor(max_workers=thread_count) as executor:
            # Submit all tasks
            futures = [executor.submit(normalize_postal_code) for _ in range(thread_count)]
            
            # Collect results
            for future in as_completed(futures, timeout=15):
                try:
                    result = future.result()
                    results.append(result)
                except Exception as e:
                    pytest.fail(f"Thread execution failed: {e}")
        
        # Verify all threads got the same result
        assert len(results) == thread_count, f"Expected {thread_count} results, got {len(results)}"
        
        for i, result in enumerate(results):
            assert result == expected_result, f"Thread {i} got unexpected result: {result}"
    
    def test_normalize_thread_safety_barrier_style(self):
        """Test thread safety using barrier-style synchronization (closer to Java version)."""
        thread_count = 50  # Reduced for faster execution
        test_postal_code = "10001-6789"
        expected_result = "10001"
        results = []
        barrier = threading.Barrier(thread_count)
        start_event = threading.Event()
        
        def worker():
            """Worker function that waits for synchronization."""
            try:
                # Wait for all threads to be ready
                start_event.wait(timeout=10)
                
                # Synchronize all threads to increase contention
                barrier.wait(timeout=10)
                
                # Perform normalization
                result = self.postal_code_attribute.normalize(test_postal_code)
                results.append(result)
                
            except Exception as e:
                pytest.fail(f"Worker thread failed: {e}")
        
        # Create and start threads
        threads = []
        for i in range(thread_count):
            thread = threading.Thread(target=worker)
            thread.start()
            threads.append(thread)
        
        # Start all threads simultaneously
        start_event.set()
        
        # Wait for all threads to complete
        for thread in threads:
            thread.join(timeout=15)
            if thread.is_alive():
                pytest.fail("Thread did not complete in time")
        
        # Verify results
        assert len(results) == thread_count
        for result in results:
            assert result == expected_result
    
    def test_normalize_should_handle_edge_cases(self):
        """Test edge cases for normalization."""
        # Test short postal codes (less than 5 characters)
        assert self.postal_code_attribute.normalize("1234 ") == "1234"
        assert self.postal_code_attribute.normalize("123") == "123"
        assert self.postal_code_attribute.normalize("12") == "12"
        assert self.postal_code_attribute.normalize("1") == "1"
        
        # Test null and empty values
        assert self.postal_code_attribute.normalize(None) is None
        assert self.postal_code_attribute.normalize("") == ""
        
        # Test exactly 5 characters
        assert self.postal_code_attribute.normalize("10001") == "10001"
        assert self.postal_code_attribute.normalize(" 10001") == "10001"
        assert self.postal_code_attribute.normalize("10001-6789") == "10001"
        assert self.postal_code_attribute.normalize("100016789") == "10001"
    
    def test_serialization(self):
        """Test serialization and deserialization."""
        # Serialize the attribute using pickle
        serialized_data = pickle.dumps(self.postal_code_attribute)
        
        # Deserialize the attribute
        deserialized_attribute = pickle.loads(serialized_data)
        
        # Test various postal code values with both original and deserialized attributes
        test_values = [
            "10001",
            "10001-6789",
            "90210-1234",
            "30301",
            "60601-2345",
            "K1A 0A7",
            "k1a0a7",
            "M5V 3L9",
            "H3Z2Y7"
        ]
        
        for value in test_values:
            assert (self.postal_code_attribute.get_name() == 
                   deserialized_attribute.get_name()), \
                   "Attribute names should match"
            
            assert (self.postal_code_attribute.get_aliases() == 
                   deserialized_attribute.get_aliases()), \
                   "Attribute aliases should match"
            
            assert (self.postal_code_attribute.normalize(value) == 
                   deserialized_attribute.normalize(value)), \
                   f"Normalization should be identical for value: {value}"
            
            assert (self.postal_code_attribute.validate(value) == 
                   deserialized_attribute.validate(value)), \
                   f"Validation should be identical for value: {value}"
    
    @pytest.mark.parametrize("input_code,expected_output", [
        ("10001-6789", "10001"),
        ("10001", "10001"),
        ("K1A0A6", "K1A 0A6"),
        ("k1a0a6", "K1A 0A6"),
        ("K1A 0A6", "K1A 0A6"),
        ("m5v3l9", "M5V 3L9"),
        ("H3Z2Y7", "H3Z 2Y7"),
        ("t2x1v4", "T2X 1V4"),
        ("95123-6789", "95123"),
        ("90210-1234", "90210"),
        ("  10001  ", "10001"),
        ("100016789", "10001"),
    ])
    def test_normalize_parametrized(self, input_code, expected_output):
        """Parametrized test for normalization with various postal codes."""
        assert self.postal_code_attribute.normalize(input_code) == expected_output
    
    @pytest.mark.parametrize("valid_code", [
        "95123 ",
        " 95123",
        "95123",
        "95123-6789",
        "65201-6789",
        "K1A 0A7",
        "K1A0A7",
        "k1a 0a7",
        "k1a0a7",
        "M5V 3L9",
        "H3Z 2Y7",
        "T2X 1V4",
        " K1A 0A7 ",
        "  K1A0A7  ",
    ])
    def test_validate_valid_codes_parametrized(self, valid_code):
        """Parametrized test for validation with valid postal codes."""
        assert self.postal_code_attribute.validate(valid_code) is True
    
    @pytest.mark.parametrize("invalid_code", [
        None,
        "",
        "1234",
        "12345",
        "54321",
        "123456",
        "1234-5678",
        "abcde",
        "K1A",
        "K1A 0A",
        "K1A 0A67",
        "K11 0A6",
        "KAA 0A6",
        "K1A 0AA",
    ])
    def test_validate_invalid_codes_parametrized(self, invalid_code):
        """Parametrized test for validation with invalid postal codes."""
        assert self.postal_code_attribute.validate(invalid_code) is False
    
    def test_normalize_us_zip_codes(self):
        """Test normalization of various US ZIP code formats."""
        # Standard 5-digit ZIP codes
        assert self.postal_code_attribute.normalize("10001") == "10001"
        assert self.postal_code_attribute.normalize("90210") == "90210"
        assert self.postal_code_attribute.normalize("30301") == "30301"
        
        # ZIP+4 format (should return first 5 digits)
        assert self.postal_code_attribute.normalize("10001-6789") == "10001"
        assert self.postal_code_attribute.normalize("90210-1234") == "90210"
        assert self.postal_code_attribute.normalize("30301-5678") == "30301"
        
        # 9-digit format without dash (should return first 5 digits)
        assert self.postal_code_attribute.normalize("100016789") == "10001"
        assert self.postal_code_attribute.normalize("902101234") == "90210"
        assert self.postal_code_attribute.normalize("303015678") == "30301"
        
        # With whitespace
        assert self.postal_code_attribute.normalize("  10001  ") == "10001"
        assert self.postal_code_attribute.normalize("  10001-6789  ") == "10001"
    
    def test_normalize_canadian_postal_codes_comprehensive(self):
        """Test comprehensive normalization of Canadian postal codes."""
        # Various formats should all normalize to A1A 1A1 format
        test_cases = [
            ("K1A0A6", "K1A 0A6"),
            ("k1a0a6", "K1A 0A6"),
            ("K1A 0A6", "K1A 0A6"),
            ("K1a 0a6", "K1A 0A6"),
            ("k1A0A6", "K1A 0A6"),
            ("m5v3l9", "M5V 3L9"),
            ("M5V3L9", "M5V 3L9"),
            ("M5v 3l9", "M5V 3L9"),
            ("H3Z2Y7", "H3Z 2Y7"),
            ("h3z2y7", "H3Z 2Y7"),
            ("t2x1v4", "T2X 1V4"),
            ("T2X1V4", "T2X 1V4"),
            ("v6b1a1", "V6B 1A1"),
            ("V6B1A1", "V6B 1A1"),
            ("n2l3g1", "N2L 3G1"),
            ("N2L3G1", "N2L 3G1"),
        ]
        
        for input_code, expected_output in test_cases:
            assert self.postal_code_attribute.normalize(input_code) == expected_output, \
                   f"Failed for input: {input_code}"
    
    def test_validate_comprehensive_us_zip_codes(self):
        """Test validation of various US ZIP code scenarios."""
        # Valid US ZIP codes (these should be valid according to the validator)
        valid_us_codes = [
            "95123",
            "95123-6789",
            "90210",
            "90210-1234",
            "10001",
            "10001-5678",
            "65201-6789",
            " 95123 ",
            "  95123-6789  ",
        ]
        
        for code in valid_us_codes:
            assert self.postal_code_attribute.validate(code) is True, \
                   f"Should be valid: {code}"
    
    def test_validate_comprehensive_canadian_postal_codes(self):
        """Test validation of various Canadian postal code scenarios."""
        # Valid Canadian postal codes
        valid_canadian_codes = [
            "K1A 0A7",  # Note: Using 0A7 instead of 0A6 to avoid placeholder
            "K1A0A7",
            "k1a 0a7",
            "k1a0a7",
            "M5V 3L9",
            "H3Z 2Y7",
            "T2X 1V4",
            "V6B 1A1",
            "N2L 3G1",
            " K1A 0A7 ",
            "  K1A0A7  ",
        ]
        
        for code in valid_canadian_codes:
            assert self.postal_code_attribute.validate(code) is True, \
                   f"Should be valid: {code}"
    
    def test_edge_cases_whitespace_handling(self):
        """Test various edge cases for whitespace handling."""
        # Leading/trailing whitespace should be trimmed
        assert self.postal_code_attribute.normalize("   10001   ") == "10001"
        assert self.postal_code_attribute.normalize("\t10001\t") == "10001"
        assert self.postal_code_attribute.normalize("\n10001\n") == "10001"
        
        # Canadian postal codes with whitespace
        assert self.postal_code_attribute.normalize("   K1A0A6   ") == "K1A 0A6"
        assert self.postal_code_attribute.normalize("\tK1A0A6\t") == "K1A 0A6"
        
        # Mixed whitespace types
        assert self.postal_code_attribute.normalize(" \t 10001 \n ") == "10001"
        assert self.postal_code_attribute.normalize(" \t K1A0A6 \n ") == "K1A 0A6"

File: lib/java/opentoken/src/test/java/com/truveta/opentoken/attributes/person/PostalCodeAttributeTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.person;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.TimeUnit;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class PostalCodeAttributeTest {
    private PostalCodeAttribute postalCodeAttribute;

    @BeforeEach
    void setUp() {
        postalCodeAttribute = new PostalCodeAttribute();
    }

    @Test
    void getName_ShouldReturnPostalCode() {
        assertEquals("PostalCode", postalCodeAttribute.getName());
    }

    @Test
    void getAliases_ShouldReturnPostalCodeAndZipCode() {
        String[] expectedAliases = { "PostalCode", "ZipCode" };
        assertArrayEquals(expectedAliases, postalCodeAttribute.getAliases());
    }

    @Test
    void normalize_ShouldReturnFirst5Digits() {
        assertEquals("10001", postalCodeAttribute.normalize("10001-6789"));
        assertEquals("10001", postalCodeAttribute.normalize("10001"));
    }

    @Test
    void normalize_ShouldHandleCanadianPostalCodes() {
        assertEquals("K1A 0A6", postalCodeAttribute.normalize("K1A0A6"));
        assertEquals("K1A 0A6", postalCodeAttribute.normalize("k1a0a6"));
        assertEquals("K1A 0A6", postalCodeAttribute.normalize("K1A 0A6"));
        assertEquals("M5V 3L9", postalCodeAttribute.normalize("m5v3l9"));
        assertEquals("H3Z 2Y7", postalCodeAttribute.normalize("H3Z2Y7"));
        assertEquals("T2X 1V4", postalCodeAttribute.normalize("t2x1v4"));
    }

    @Test
    void validate_ShouldReturnTrueForValidPostalCodes() {
        assertTrue(postalCodeAttribute.validate("95123 "));
        assertTrue(postalCodeAttribute.validate(" 95123"));
        assertTrue(postalCodeAttribute.validate("95123"));
        assertTrue(postalCodeAttribute.validate("95123-6789"));
        assertTrue(postalCodeAttribute.validate("65201-6789"));
    }

    @Test
    void validate_ShouldReturnTrueForValidCanadianPostalCodes() {
        assertTrue(postalCodeAttribute.validate("K1A 0A7"));
        assertTrue(postalCodeAttribute.validate("K1A0A7"));
        assertTrue(postalCodeAttribute.validate("k1a 0a7"));
        assertTrue(postalCodeAttribute.validate("k1a0a7"));
        assertTrue(postalCodeAttribute.validate("M5V 3L9"));
        assertTrue(postalCodeAttribute.validate("H3Z 2Y7"));
        assertTrue(postalCodeAttribute.validate("T2X 1V4"));
        assertTrue(postalCodeAttribute.validate(" K1A 0A7 "));
        assertTrue(postalCodeAttribute.validate("  K1A0A7  "));
    }

    @Test
    void normalize_ShouldHandleWhitespace() {
        PostalCodeAttribute attribute = new PostalCodeAttribute();

        // Test different types of whitespace
        assertEquals("10001", attribute.normalize("10001"), "No whitespace");
        assertEquals("10001", attribute.normalize(" 10001"), "Leading space");
        assertEquals("10001", attribute.normalize("10001 "), "Trailing space");
        assertEquals("10001", attribute.normalize(" 10001 "), "Leading and trailing spaces");
        assertEquals("10   001", attribute.normalize("  10   001  "), "Multiple spaces");
    }

    @Test
    void normalize_ShouldNotHandleInnerWhitespace() {
        PostalCodeAttribute attribute = new PostalCodeAttribute();

        // Test inner whitespace
        assertEquals("1 0 0 0 1", attribute.normalize("1 0 0 0 1"), "Spaces between digits");
        assertEquals("10\t001", attribute.normalize("10\t001"), "Tab character");
        assertEquals("10\n001", attribute.normalize("10\n001"), "Newline character");
        assertEquals("10\r\n001", attribute.normalize("10\r\n001"), "Carriage return and newline");
        assertEquals("K1A  0A7", attribute.normalize("K1A  0A7"),
                "Inner space should not be normalized for Canadian postal code");
    }

    @Test
    void validate_ShouldReturnFalseForInvalidPostalCodes() {
        assertFalse(postalCodeAttribute.validate(null), "Null value should not be allowed");
        assertFalse(postalCodeAttribute.validate(""), "Empty value should not be allowed");
        assertFalse(postalCodeAttribute.validate("1234"), "Short postal code should not be allowed");
        assertFalse(postalCodeAttribute.validate("12345"), "Invalid postal code should not be allowed");
        assertFalse(postalCodeAttribute.validate("54321"), "Invalid postal code should not be allowed");
        assertFalse(postalCodeAttribute.validate("123456"), "Long postal code should not be allowed");
        assertFalse(postalCodeAttribute.validate("1234-5678"), "Invalid format should not be allowed");
        assertFalse(postalCodeAttribute.validate("abcde"), "Non-numeric should not be allowed");

        // Invalid Canadian postal code formats
        assertFalse(postalCodeAttribute.validate("K1A"), "Incomplete Canadian postal code should not be allowed");
        assertFalse(postalCodeAttribute.validate("K1A 0A"), "Incomplete Canadian postal code should not be allowed");
        assertFalse(postalCodeAttribute.validate("K1A 0A67"), "Too long Canadian postal code should not be allowed");
        assertFalse(postalCodeAttribute.validate("K11 0A6"),
                "Invalid Canadian postal code format should not be allowed");
        assertFalse(postalCodeAttribute.validate("KAA 0A6"),
                "Invalid Canadian postal code format should not be allowed");
        assertFalse(postalCodeAttribute.validate("K1A 0AA"),
                "Invalid Canadian postal code format should not be allowed");
    }

    @Test
    void normalize_ThreadSafety() throws InterruptedException {
        final int threadCount = 100;
        final CountDownLatch startLatch = new CountDownLatch(1);
        final CountDownLatch finishLatch = new CountDownLatch(threadCount);
        final CyclicBarrier barrier = new CyclicBarrier(threadCount);
        final List<String> results = Collections.synchronizedList(new ArrayList<>());
        final String testPostalCode = "10001-6789";

        for (int i = 0; i < threadCount; i++) {
            Thread thread = new Thread(() -> {
                try {
                    startLatch.await(); // Wait for all threads to be ready
                    barrier.await(); // Synchronize to increase contention

                    // Perform normalization
                    String result = postalCodeAttribute.normalize(testPostalCode);
                    results.add(result);

                    finishLatch.countDown();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
            thread.start();
        }

        startLatch.countDown(); // Start all threads
        finishLatch.await(15, TimeUnit.SECONDS); // Wait for all threads to complete

        // Verify all threads got the same result
        assertEquals(threadCount, results.size());
        for (String result : results) {
            assertEquals("10001", result);
        }
    }

    @Test
    void normalize_ShouldHandleEdgeCases() {
        // Test short postal codes (less than 5 characters)
        assertEquals("1234", postalCodeAttribute.normalize("1234 "));
        assertEquals("123", postalCodeAttribute.normalize("123"));
        assertEquals("12", postalCodeAttribute.normalize("12"));
        assertEquals("1", postalCodeAttribute.normalize("1"));

        // Test null and empty values
        assertNull(postalCodeAttribute.normalize(null));
        assertEquals("", postalCodeAttribute.normalize(""));

        // Test exactly 5 characters
        assertEquals("10001", postalCodeAttribute.normalize("10001"));

        assertEquals("10001", postalCodeAttribute.normalize(" 10001"));

        assertEquals("10001", postalCodeAttribute.normalize("10001-6789"));
        assertEquals("10001", postalCodeAttribute.normalize("100016789"));
    }

    @Test
    void testSerialization() throws Exception {
        // Serialize the attribute
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ObjectOutputStream oos = new ObjectOutputStream(baos);
        oos.writeObject(postalCodeAttribute);
        oos.close();

        // Deserialize the attribute
        ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
        ObjectInputStream ois = new ObjectInputStream(bais);
        PostalCodeAttribute deserializedAttribute = (PostalCodeAttribute) ois.readObject();
        ois.close();

        // Test various postal code values with both original and deserialized
        // attributes
        String[] testValues = {
                "10001",
                "10001-6789",
                "90210-1234",
                "30301",
                "60601-2345",
                "K1A 0A7",
                "k1a0a7",
                "M5V 3L9",
                "H3Z2Y7"
        };

        for (String value : testValues) {
            assertEquals(
                    postalCodeAttribute.getName(),
                    deserializedAttribute.getName(),
                    "Attribute names should match");

            assertArrayEquals(
                    postalCodeAttribute.getAliases(),
                    deserializedAttribute.getAliases(),
                    "Attribute aliases should match");

            assertEquals(
                    postalCodeAttribute.normalize(value),
                    deserializedAttribute.normalize(value),
                    "Normalization should be identical for value: " + value);

            assertEquals(
                    postalCodeAttribute.validate(value),
                    deserializedAttribute.validate(value),
                    "Validation should be identical for value: " + value);
        }
    }
}
